File: graveler_lock_py_cpp/graveler_lock_2.py
# This script was written mostly to demonstrate the inefficiency of
# branching in python. The reason for this is that the CPU tries to
# read ahead which it cannot do or only worse at a branch point.
# Furthermore the script makes use of the built in functions of the
# random library.
# This leads to a speed up of a factor ~3

# Performance on 1M: ~30sec

import random
import time

n = 1000000
d4 = (1, 0, 0, 0)
k=231

start_time = time.time()

max_value = 0
for _ in range(n):
    value = sum(random.choices(d4, k=k))
    max_value = max(max_value, value)

total_time = time.time() - start_time

print(f"Highest Ones Roll: {max_value}")
print(f"Number of Roll Sessions: {n}")
print(f"Total Elapsed Time: {round(total_time, 1)}s")

File: graveler_lock_py_cpp/graveler_lock_2.cpp
// A better random number generator can lead to significan improvements.
// Simply using the Mersenne Twister results in a speed up by almost
// a factor of 10. This shows in particular that the random number generation
// is the bottleneck of the script.

// Performance on 1M: ~0.4sec
// Performance on 100M: ~40sec

#include <iostream>
#include <random>
#include <chrono>

int main() {
    int n = 1e8;
    int k = 231;

    auto start_time = std::chrono::high_resolution_clock::now();

    std::mt19937 rng(std::random_device{}()); // Mersenne Twister RNG
    std::uniform_int_distribution<int> dist(0, 3);

    int max_value = 0;
    for (int i = 0; i < n; ++i) {
        int value = 0;
        for (int j = 0; j < k; ++j) {
            value += (dist(rng) == 0);
        }
        if (value > max_value) {
            max_value = value;
        }
    }

    auto total_time = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now() - start_time);

    std::cout << "Highest Ones Roll: " << max_value << std::endl;
    std::cout << "Number of Roll Sessions: " << n << std::endl;
    std::cout << "Total Elapsed Time: " << total_time.count() * 1e-3 << "s" << std::endl;
    
    return 0;
}